chia/server/ws_connection.py

```diff
@@ -391,34 +400,135 @@ async def _wait_and_retry(self, msg: Message, queue: asyncio.Queue):
-                    f"Rate limiting ourselves. message type: {ProtocolMessageTypes(message.type).name}, "
-                    f"peer: {self.peer_host}"
+                    f"Rate limiting ourselves. message type: {message_type.name}{more_info}, peer: {self.peer_host}"
                 )
 
                 # TODO: fix this special case. This function has rate limits which are too low.
-                if ProtocolMessageTypes(message.type) != ProtocolMessageTypes.respond_peers:
-                    asyncio.create_task(self._wait_and_retry(message, self.outgoing_queue))
+                if message_type != ProtocolMessageTypes.respond_peers:
+                    asyncio.create_task(self._wait_and_retry(message_to_send, self.outgoing_queue))
 
                 return None
             else:
                 self.log.debug(
-                    f"Not rate limiting ourselves. message type: {ProtocolMessageTypes(message.type).name}, "
+                    f"Not rate limiting ourselves. message type: {message_type.name}{more_info}, "
                     f"peer: {self.peer_host}"
                 )
 
         await self.ws.send_bytes(encoded)
-        self.log.debug(f"-> {ProtocolMessageTypes(message.type).name} to peer {self.peer_host} {self.peer_node_id}")
+        self.log.debug(f"-> {message_type.name} to peer {self.peer_host} {self.peer_node_id}")
         self.bytes_written += size
 
+    async def _decompress_message(self, full_message_loaded: Message, test_mode: bool = False) -> Optional[Message]:
+        # This method is only called for ProtocolMessageTypes.wrapped_compressed
+        # The 'test_mode' is only to make testing easier
+        try:
+            if Capability.CAN_DECOMPRESS_MESSAGES in self.local_capabilities and len(full_message_loaded.data) > 0:
+                # Check the uncompressed size before doing the decompression (must be less than 4 GiB)
+                decompressed_size = get_decompressed_size(full_message_loaded.data[1:])
+                self.log.debug(
+                    f"Received compressed message {ProtocolMessageTypes(full_message_loaded.data[0]).name}: "
+                    f"compressed {len(full_message_loaded.data)}, decompressed {decompressed_size}"
+                )
+                if decompressed_size > 0 and decompressed_size < (2 ** (LENGTH_BYTES * 8)):
+                    # Replace the message so it appears as if it was sent uncompressed
+                    # The rate limiter has already checked the message (using the size of the compressed message)
+                    full_message_loaded = Message(
+                        uint8(full_message_loaded.data[0]),
+                        full_message_loaded.id,
+                        zstd.decompress(full_message_loaded.data[1:]),
+                    )
+                    return full_message_loaded
+                else:
+                    self.log.debug(" -> It is either not using Zstandard or it is too big")
```

Comment:
I believe this should be a disconnect and probably a ban

Reply:
Yes, disconnect and ban is a valid outcome.